msg/templates/uorb_microcdr/microRTPS_client.cpp.em

```diff
@@ -76,184 +76,233 @@ receive_base_types = [s.short_name for idx, s in enumerate(spec) if scope[idx] =
+
+					if (0 < (read = transport_node->write(static_cast<char>(@(rtps_message_id(ids, topic))), data_buffer, length))) {
+						data->total_sent += read;
+						tx_last_sec_read += read;
+						++data->sent;
+					}
+
+@[        if topic == 'Timesync' or topic == 'timesync']@
+				}
+
+@[        end if]@
+			}
+		}
+@[    end for]@
 
-    struct timespec end;
-    px4_clock_gettime(CLOCK_REALTIME, &end);
-    double elapsed_secs = end.tv_sec - begin.tv_sec + (end.tv_nsec - begin.tv_nsec) / 1e9;
-    PX4_INFO("SENT: %" PRIu64 " messages in %d LOOPS, %" PRIu64 " bytes in %.03f seconds - %.02fKB/s",
-                sent, loop, total_sent, elapsed_secs, total_sent / (1e3 * elapsed_secs));
+		if (hrt_absolute_time() - last_stats_update >= 1_s) {
+			data->sent_last_sec = tx_last_sec_read;
+			if (data->datarate > 0) {
+				bandwidth_mult = static_cast<float>(data->datarate) / static_cast<float>(tx_last_sec_read);
+				// Apply a low-pass filter to determine the new TX interval
+				tx_interval += 0.5f * (tx_interval / bandwidth_mult - tx_interval);
+				// Clamp the interval between 1 and 1000 ms
+				tx_interval = math::constrain(tx_interval, MIN_TX_INTERVAL_US, MAX_TX_INTERVAL_US);
+			}
+			tx_last_sec_read = 0;
+			last_stats_update = hrt_absolute_time();
+		}
 
-    delete subs;
+		px4_usleep(tx_interval);
 
-    return nullptr;
+		++data->sent_loop;
+	}
+
+	delete(data);
+	delete(subs);
+
+	return nullptr;
 }
 
-static int launch_send_thread(pthread_t &sender_thread)
+static int launch_send_thread(pthread_t &sender_thread, struct SendThreadArgs &args)
 {
-    pthread_attr_t sender_thread_attr;
-    pthread_attr_init(&sender_thread_attr);
-    pthread_attr_setstacksize(&sender_thread_attr, PX4_STACK_ADJUSTED(2250));
-    struct sched_param param;
-    (void)pthread_attr_getschedparam(&sender_thread_attr, &param);
-    param.sched_priority = SCHED_PRIORITY_DEFAULT;
-    (void)pthread_attr_setschedparam(&sender_thread_attr, &param);
-    pthread_create(&sender_thread, &sender_thread_attr, send, nullptr);
-    if (pthread_setname_np(sender_thread, "micrortps_client_send"))
-    {
-        PX4_ERR("Could not set pthread name (%d)", errno);
-    }
-    pthread_attr_destroy(&sender_thread_attr);
-
-    return 0;
+	pthread_attr_t sender_thread_attr;
+	pthread_attr_init(&sender_thread_attr);
+	pthread_attr_setstacksize(&sender_thread_attr, PX4_STACK_ADJUSTED(2250));
+	struct sched_param param;
+	(void)pthread_attr_getschedparam(&sender_thread_attr, &param);
+	param.sched_priority = SCHED_PRIORITY_DEFAULT;
+	(void)pthread_attr_setschedparam(&sender_thread_attr, &param);
+	int rc = pthread_create(&sender_thread, &sender_thread_attr, &send, (void *)&args);
+	if (rc != 0) {
+		errno = rc;
+		PX4_ERR("Could not create send thread (%d)", errno);
+		return -1;
+	}
+	rc = pthread_setname_np(sender_thread, "urtpsclient_snd");
+	if (pthread_setname_np(sender_thread, "urtpsclient_snd")) {
+		errno = rc;
```

Comment:
Why do we (mis)use `errno` here instead of just returning it as the result?

Reply:
well if I print errno, it doesn't provide the result of the np. I guess I could just print rc, just not sure if that should be done that way.